### PR TITLE
segment-region: only ignore if candidate outside Border

### LINKED ISSUE #139: Processor terminates with exception: ocrd-tesserocr-segment-region exited with non-zero
Physical page 129 of http://nbn-resolving.de/urn:nbn:de:bsz:180-digad-10044 raises a fatal exception in the standard [workflow for slower processors](https://ocr-d.de/en/workflows):

    20:13:52.948 INFO processor.TesserocrSegmentRegion - INPUT FILE 128 / IMG_OCR-D-BIN-DENOISE-DESKEW_460017
    20:13:52.971 INFO processor.TesserocrSegmentRegion - Page 'IMG_OCR-D-BIN-DENOISE-DESKEW_460017' images will use 300 DPI from image meta-data
    20:13:52.971 INFO processor.TesserocrSegmentRegion - Detecting regions in page 'IMG_OCR-D-BIN-DENOISE-DESKEW_460017'
    20:13:53.352 INFO processor.TesserocrSegmentRegion - Detected region 'region0000': 1425,976 1421,976 1424,2217 1425,2217 (HORZ_LINE)
    20:13:53.353 INFO processor.TesserocrSegmentRegion - Detected region 'region0001': 1425,1788 1423,976 1413,976 1416,2194 1425,2194 (HORZ_LINE)
    20:13:53.354 INFO processor.TesserocrSegmentRegion - Detected region 'region0002': 523,97 525,967 958,966 956,97 (TABLE)
    20:13:53.354 INFO processor.TesserocrSegmentRegion - Detected region 'region0003': 392,292 432,292 434,1160 394,1160 (FLOWING_TEXT)
    20:13:53.355 INFO processor.TesserocrSegmentRegion - Detected region 'region0004': 378,162 401,162 401,307 378,307 (FLOWING_TEXT)
    20:13:53.355 INFO processor.TesserocrSegmentRegion - Detected region 'region0005': 92,110 330,109 332,1157 94,1158 (TABLE)
    Traceback (most recent call last):
      File "/home/stweil/src/github/OCR-D/venv-20200815/bin/ocrd-tesserocr-segment-region", line 8, in <module>
        sys.exit(ocrd_tesserocr_segment_region())
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/click/core.py", line 829, in __call__
        return self.main(*args, **kwargs)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/click/core.py", line 782, in main
        rv = self.invoke(ctx)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
        return ctx.invoke(self.callback, **ctx.params)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/click/core.py", line 610, in invoke
        return callback(*args, **kwargs)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/ocrd_tesserocr/cli.py", line 16, in ocrd_tesserocr_segment_region
        return ocrd_cli_wrap_processor(TesserocrSegmentRegion, *args, **kwargs)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/ocrd/decorators.py", line 102, in ocrd_cli_wrap_processor
        run_processor(processorClass, ocrd_tool, mets, workspace=workspace, **kwargs)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/ocrd/processor/base.py", line 61, in run_processor
        processor.process()
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/ocrd_tesserocr/segment_region.py", line 168, in process
        self._process_page(layout, page, page_image, page_coords, input_file.pageId)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/ocrd_tesserocr/segment_region.py", line 221, in _process_page
        polygon = polygon_for_parent(polygon, page)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/ocrd_tesserocr/segment_region.py", line 327, in polygon_for_parent
        raise Exception("intersection of would-be segment with parent is empty")
    Exception: intersection of would-be segment with parent is empty
    Traceback (most recent call last):
      File "/home/stweil/src/github/OCR-D/venv-20200815/bin/ocrd", line 8, in <module>
        sys.exit(cli())
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/click/core.py", line 829, in __call__
        return self.main(*args, **kwargs)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/click/core.py", line 782, in main
        rv = self.invoke(ctx)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
        return _process_result(sub_ctx.command.invoke(sub_ctx))
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
        return ctx.invoke(self.callback, **ctx.params)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/click/core.py", line 610, in invoke
        return callback(*args, **kwargs)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/ocrd/cli/process.py", line 27, in process_cli
        run_tasks(mets, log_level, page_id, tasks, overwrite)
      File "/home/stweil/src/github/OCR-D/venv-20200815/lib/python3.7/site-packages/ocrd/task_sequence.py", line 148, in run_tasks
        raise Exception("%s exited with non-zero return value %s" % (task.executable, returncode))
    Exception: ocrd-tesserocr-segment-region exited with non-zero return value 1

### PR DESCRIPTION
**Pull request by bertsky**
Fix #139

### CODE REVIEW

**Comment by bertsky** [comment]
@kba if you want, I can export the last 3 commits as extra PR.

File: .circleci/config.yml [hunk 1]
```diff
@@ -6,11 +6,12 @@ jobs:
 
   build-python36:
     docker:
-      - image: ubuntu:18.04
+      - image: ocrd/core
     steps:
       - run: apt-get update && apt-get install -y --no-install-recommends make git curl
       - checkout
-      - run: make deps-ubuntu deps-test deps install repo/assets
+      - run: make deps-ubuntu
```
Comment (kba) [inline .circleci/config.yml:13]: What about `deps-test`? Where will CI get pytest from?

File: Makefile [hunk 2]
```diff
@@ -25,14 +25,14 @@ help:
 	@echo "                   from Alexander Pozdnyakov which provides 4.1.0."
 	@echo "                   See https://launchpad.net/~alex-p/+archive/ubuntu/tesseract-ocr"
 	@echo "                   for details.)"
-	@echo "    deps          Install python deps via pip"
-	@echo "    deps-test     Install testing python deps via pip"
-	@echo "    install       Install"
+	@echo "    deps          Install Python deps for install via pip"
```
Comment (kba) [inline Makefile:28]: If you add those descriptions as comments directly above the targets, you can run [`shinclude -i Makefile`](https://github.com/kba/shinclude) which will in turn (see the `BEGIN-EVAL` comment above) run [makefile-parser](https://github.com/kba/makefile-parser) to generate the help.

File: Makefile [hunk 3]
```diff
@@ -110,6 +110,8 @@ test-cli: test/assets
 #
 
 # Clone OCR-D/assets to ./repo/assets
+# FIXME does not work if already checked out
+# FIXME should be a proper (VCed) submodule
```
Comment (kba) [inline Makefile:114]: true, we're using submodules in most places now.

File: ocrd_tesserocr/segment_region.py [hunk 4]
```diff
@@ -219,9 +219,15 @@ def _process_page(self, it, page, page_image, page_coords, page_id):
             else:
                 polygon = polygon_from_x0y0x1y1(bbox)
             polygon = coordinates_for_segment(polygon, page_image, page_coords)
-            polygon = polygon_for_parent(polygon, page)
+            polygon2 = polygon_for_parent(polygon, page)
+            if polygon2 is not None:
+                polygon = polygon2
             points = points_from_polygon(polygon)
             coords = CoordsType(points=points)
+            if polygon2 is None:
```
Comment (kba) [inline ocrd_tesserocr/segment_region.py:227]: Why the `is None` check after the two lines before? Can't this be combined with the `if polygon2 is not None` check above?

File: .circleci/config.yml (hunk 1 above)
Comment (kba) [reply .circleci/config.yml:13]: Oh wait, you made test depend on deps-test, right?

File: Makefile (hunk 2 above)
Comment (bertsky) [reply Makefile:28]: Oh, I didn't know you were doing it like this!

And the order of the help comments is the order of the targets?

File: ocrd_tesserocr/segment_region.py (hunk 4 above)
Comment (bertsky) [reply ocrd_tesserocr/segment_region.py:227]: Because we also want to generate an informative log message (which needs `points`).

Comment (kba) [reply ocrd_tesserocr/segment_region.py:227]: I see, makes sense.

File: Makefile (hunk 2 above)
Comment (bertsky) [reply Makefile:28]: So do you want me to rewrite those comments (and the order) of the rules?

Comment (kba) [reply Makefile:28]: Yes, makefile-parser parses a subset of Makefile syntax (assignments and targets - does that remind you of somehting :-P) and pairs them with the immediately preceding comment and then outputs, in document order, help for targets and variables.

File: .circleci/config.yml (hunk 1 above)
Comment (bertsky) [reply .circleci/config.yml:13]: Yes, exactly. It seemed more natural to encapsulate all these details. And the cost of running `pip` each time before the test seemed worth it...

File: Makefile (hunk 2 above)
Comment (kba) [reply Makefile:28]: It's up to you, but if you want to see how it is being generated by shinclude/makefile-parser, see https://github.com/OCR-D/ocrd_tesserocr/commit/c2f3ad84b3f47008a128afabd95cba71c116b4ac

Comment (bertsky) [reply Makefile:28]: Wow, that's quite an undertaking! Does using regexps to partially parse the makefile allow you to check for unsupported syntax or traps, so you can warn the user?

What about `FIXME` comments and the like?

Comment (kba) [reply Makefile:28]: > allow you to check for unsupported syntax or traps,

In a way yes, there will be warnings about `UNHANDLED LINE` if no regex matches (for many constructs, like conditionals and multiline variable-assignment.

> `FIXME` comments

I deleted those manually I have to admit :)

Comment (bertsky) [reply Makefile:28]: 9e6ad41 should suffice

File: Makefile (hunk 3 above)
Comment (bertsky) [reply Makefile:121]: Let's address this another time.

**Pull request merged**